File: Baekjoon_Solve/BFS/16988.java
// Baaaaaaaaaduk2 (Easy) (16988번)
///////////////////////////////////////////////////////////////////////////////////////////////
  // 문제: https://www.acmicpc.net/problem/16988
  // BFS
  // 이 문제를 풀면서 자바를 더 공부해야겠다는 생각이 든다.
  // 객체의 복사(얕은 복사, 깊은 복사)에 대해 공부해야겠다.
  // 파이썬에서는 그냥 잘 됐던 것 같은데, 왜 이게 얕은 복사가 되어서 재귀가 끝났을 때 ArrayList가 초기화되는건지 모르겠다...
  // 아직 자바는 기초가 많이 약하다
///////////////////////////////////////////////////////////////////////////////////////////////
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static class Pair {
        int y;
        int x;

        Pair(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }

    static void comb(int k, int s, Stack<Pair> arr) {
        if (k == 2) {
            Stack<Pair> copied = new Stack<>();
            copied.add(arr.get(0));
            copied.add(arr.get(1));
            cmblst.add(copied); // 그냥 arr를 add 해주니까 얕은 복사가 되어 재귀호출 종료 후 cmblst가 초기화됨
            return;
        }
        for (int i=s; i<K; i++) {
            arr.add(cand.get(i));
            comb(k + 1, i + 1, arr);
            arr.pop();
        }
    }


    static int bfs() {
        int TOTAL = 0;
        boolean[][] visited = new boolean[N][M];
        for (int i=0; i<N; i++) {
            for (int j=0; j<M; j++) {

                if (graph[i][j] == 2 && !visited[i][j]) {
                    visited[i][j] = true;
                    Queue<Pair> queue = new LinkedList<>();
                    queue.add(new Pair(i, j));
                    boolean escapable = false;
                    int cnt = 1;

                    while (!queue.isEmpty()) {
                        Pair pair = queue.poll();
                        int y = pair.y;
                        int x = pair.x;
                        for (int k=0; k<4; k++) {
                            int ny = y + dy[k];
                            int nx = x + dx[k];
                            if (0 <= ny && ny < N && 0 <= nx && nx < M && !visited[ny][nx]) {
                                if (graph[ny][nx] == 0) {
                                    escapable = true;
                                } else if (graph[ny][nx] == 2) {
                                    queue.add(new Pair(ny, nx));
                                    visited[ny][nx] = true;
                                    cnt += 1;
                                }
                            }
                        }
                    }

                    if (!escapable) {
                        TOTAL += cnt;
                    }
                }
            }
        }

        return TOTAL;
    }

    static int[][] graph;
    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static int N;
    static int M;
    static int K;
    static ArrayList<Pair> cand = new ArrayList<>(); // 0을 넣을 수 있는 후보군
    static ArrayList<Stack<Pair>> cmblst = new ArrayList<>(); // 조합 리스트

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        graph = new int[N][M];

        for (int i=0; i<N; i++) {
            String[] tmp = br.readLine().split(" ");
            for (int j=0; j<M; j++) {
                graph[i][j] = Integer.parseInt(tmp[j]);
                if (graph[i][j] == 0) {
                    cand.add(new Pair(i, j));
                }
            }
        }

        K = cand.size();

        comb(0, 0, new Stack<>());

        int max_catch = 0;
        int L = cmblst.size();
        for (int l=0; l<L; l++) {
            Stack<Pair> lst = cmblst.get(l);
            Pair pair1 = lst.pop();
            Pair pair2 = lst.pop();
            int y1 = pair1.y;
            int x1 = pair1.x;
            int y2 = pair2.y;
            int x2 = pair2.x;
            graph[y1][x1] = 1;
            graph[y2][x2] = 1;
            int ret = bfs();
            max_catch = Math.max(max_catch, ret);
            graph[y1][x1] = 0;
            graph[y2][x2] = 0;
        }

        System.out.println(max_catch);
        br.close();
    }
}

File: Baekjoon_Solve/Graph/16988.py
# Baaaaaaaaaduk2 (Easy) (16988번)
############################################################################
    # 문제: https://www.acmicpc.net/problem/16988
    # BFS
    # 0인 지점 2곳을 골라서 브루트포스로 BFS를 돌리면 된다.
    # 계속 시간초과가 나서 보니까 배열을 쓸데없이 늘리고 있었고, 쓸데없는 곳을 방문하고 있었다.
    # 그리고 nx의 범위를 0 <= nx < N으로 잡아버리는 바람에 자꾸 틀렸습니다를 받았다.
    # 생각보다는 단순한 문제였는데 너무 많이 되돌아갔던 것 같다.
    # 다른 사람의 효율적인 풀이를 보고 리팩토링했는데, deque를 쓰는 것보다 그냥 배열(스택)을 쓰는 게 더 빨랐다.
    # pypy3 기준으로 deque의 경우 1312ms, 리스트의 경우 720ms를 받았다
    # 어차피 최단 거리를 찾는 문제가 아니므로 굳이 deque를 쓸 필요는 없다.
############################################################################
import sys
from collections import deque
from itertools import combinations
input = sys.stdin.readline

dydx = [(-1, 0), (0, 1), (1, 0), (0, -1)]

N, M = map(int, input().split())
graph, lst = [], []
for i in range(N):
    graph.append(list(map(int, input().split())))
    for j in range(M):
        if not graph[i][j]:
            lst.append((i, j))


# bfs
def bfs():
    visited = [[False for _ in range(M)] for _ in range(N)]
    E = 0 # 잡히는 돌의 수
    for i in range(N):
        for j in range(M):
            if graph[i][j] == 2 and not visited[i][j]:
                visited[i][j] = True
                q = deque([(i, j)])
                escapable = False # 탈출 가능 여부
                cnt = 1
                while q:
                    y, x = q.popleft()
                    for dy, dx in dydx:
                        ny, nx = y + dy, x + dx
                        if 0 <= ny < N and 0 <= nx < M and not visited[ny][nx]:
                            if not graph[ny][nx]:
                                escapable = True
                            if graph[ny][nx] == 2:
                                q.append((ny, nx))
                                visited[ny][nx] = True
                                cnt += 1
                if not escapable: # 탈출 불가능할 경우
                    E += cnt
    return E


max_catch = 0
for cmb in list(combinations(lst, 2)):
    y1, x1 = cmb[0]
    y2, x2 = cmb[1]
    graph[y1][x1] = 1
    graph[y2][x2] = 1
    max_catch = max(max_catch, bfs())
    graph[y1][x1] = 0
    graph[y2][x2] = 0

print(max_catch)
